ads/google/utils.js

```diff
@@ -14,16 +14,289 @@
+    // User time zone offset in minutes.
+    `&u_tz=${-(new Date).getTimezoneOffset()}` +
+    (screen ? `&u_w=${screen.width}&u_h=${screen.height}` +
+              `&u_aw=${screen.availWidth}&u_ah=${screen.availHeight}` +
+              `&u_cd=${screen.colorDepth}` : '') +
+    // Font and size. These require (forbidden) getComputedStyle.
+    // `&dff=${}&dfs=16` +
+    // Ad slot's x and y coordinates. (slot is relative to viewport.)
+    (viewport ? `&adx=${slot.x + viewport.x}&ady=${slot.y + viewport.y}` +
+                // Viewport width and height.
+                `&biw=${viewport.width}&bih=${viewport.height}` :
+                `&adx=${SIZE_EXCEPTION_ID}&ady=${SIZE_EXCEPTION_ID}` +
+                `&biw=${SIZE_EXCEPTION_ID}&bih=${SIZE_EXCEPTION_ID}`) +
+    // Have to think about how to handle publisher window.
+    // `&isw=300&ish=200` +
+    // Requires looking outside in the DOM.
+    // `&ifk=3286625729` +
+    // Resend count.
+    // `&rx=0` + `
+    // Visibility state.
+    `&vis=${visNum}` +
+    // Looks at parents, needs getComputedStyle.
+    // `&pfx=0` +
+    // Browser supports SVG?
+    (global.SVGElement && global.document.createElementNS ? '&bc=1' : '') +
+    // Unique ID within window.
+    `&ifi=${slotNumber}` +
+    // Experiment IDs, if any.
+    (experimentIds ? `&eid=${experimentIds}` : '');
+
+  let truncated = false;
+  const addParameter = (name, value) => {
+    if (value != undefined) {
+      const newParam = `&${name}=${encodeURIComponent(value)}`;
+      if (url.length + newParam.length <= MAX_URL_LENGTH_) {
+        url += newParam;
+      } else {
+        truncated = true;
+      }
+    }
+  };
+
+  const addTruncatableParameter = (name, value, minUsefulLength) => {
+    if (value != undefined) {
+      const newParam = `&${name}=${encodeURIComponent(value)}`;
+      const urlLength = url.length + newParam.length;
+      if (urlLength <= MAX_URL_LENGTH_) {
+        url += newParam;
+      } else {
+        // Truncate to fit.
+        url = url.substr(
+          0, MAX_URL_LENGTH_ - TRUNC_PARAM_.length - url.length);
+        // If there is a '%' or '%\w' at the end of the URL then it's likely we've
+        // cut an escaped character (e.g. %3B = ;) so remove the whole thing
+        // rather than leave it half dangling.
+        const truncatedParam = newParam.replace(/%\w?$/, '');
+        if (truncatedParam.length >= minUsefulLength) {
+          url += truncatedParam;
+        }
+        truncated = true;
+      }
+    }
+  };
+
+  // Is this a test?
+  if (adtest != null) {
+    addParameter('adtest', data['adtest']);
+  }
+  addParameter('u_hist', getHistoryLength_(global));
+  addParameter('host', data['adHost']);
+  addParameter('to', data['tagOrigin']);
+
+  // The URL to which the ads will be targeted.
+  addTruncatableParameter('url', canonicalUrl, MIN_USEFUL_URL_PARAM_LENGTH_);
+
+  addParameter('dtd', elapsedTimeWithCeiling(timer.now(), startTime));
```

Comment:
Does `MAX_URL_LENGTH_` compensate for the `&dtd=1000000&&trunc=1` that may be added after we've already truncated?